src/metabase/api/api_key.clj

```diff
@@ -0,0 +1,58 @@
+(ns metabase.api.api-key
+  "/api/api-key endpoints for CRUD management of API Keys"
+  (:require
+   [compojure.core :refer [POST GET]]
+   [crypto.random :as crypto-random]
+   [metabase.api.common :as api]
+   [metabase.models.api-key :as api-key]
+   [metabase.models.permissions-group :as perms-group]
+   [metabase.models.user :as user]
+   [metabase.util :as u]
+   [metabase.util.i18n :refer [tru]]
+   [metabase.util.malli.schema :as ms]
+   [toucan2.core :as t2]))
+
+(defn- key-with-unique-prefix []
+  (u/auto-retry 5
+   (let [api-key (api-key/generate-key)
+         prefix (api-key/prefix api-key)]
+     ;; we could make this more efficient by generating 5 API keys up front and doing one select to remove any
+     ;; duplicates. But a duplicate should be rare enough to just do multiple queries for now.
+     (if-not (t2/exists? :model/ApiKey :key_prefix prefix)
+       api-key
+       (throw (ex-info (tru "could not generate key with unique prefix") {}))))))
+
+(api/defendpoint POST "/"
+  "Create a new API key (and an associated `User`) with the provided name and group ID."
+  [:as {{:keys [group_id name] :as _body} :body}]
+  {group_id ms/PositiveInt
+   name     ms/NonBlankString}
+  (api/check-superuser)
+  (api/checkp (not (t2/exists? :model/ApiKey :name name))
+    "name" "An API key with this name already exists.")
+  (let [api-key (key-with-unique-prefix)
+        email   (format "api-key-user-%s@api-key.invalid" name)]
+    (t2/with-transaction [_conn]
+      (let [user (first (t2/insert-returning-instances! :model/User
+                                                        {:email    email
+                                                         :password (crypto-random/base64 16)
+                                                         :type     :api-key}))]
```

Comment:
I saw that and was tempted, but the function is private and I wasn't sure whether it was worth making it public for this. WDYT?

Reply:
Oh you're right. I think that one would be fine to make public though.

Reply:
Lots of private functions, especially in model namespaces, are just private by default since they're not used outside of the namespace. But unless it's a namespace with a clearly marked public API, lots of functions are fine to make public if it's useful. This case definitely seems pretty benign to me.